src/main/java/org/xstefank/model/PersistentList.java

```diff
@@ -5,57 +5,157 @@
+        return true;
+    }
+
+    @Override
+    public void clear() {
+        super.clear();
+        if (!file.exists()) {
             return;
         }
-        list.add(username);
-        try (FileWriter fw = new FileWriter(file, true)) {
-            fw.write(username + "\n");
+        try (FileWriter fw = new FileWriter(file)){
+            fw.write("");
         } catch (IOException e) {
-            throw new IllegalStateException("Cannot write user to file", e);
+            log.error(e);
+            loadFileContent(file);
         }
     }
 
-    public boolean hasUsername(String  username) {
-        return list.contains(username);
+    @Override
+    public boolean addAll(Collection<? extends String> c) {
+        if (!super.addAll(c)) {
+            return false;
+        }
+        try (PrintWriter pw = new PrintWriter(new FileWriter(file, true))) {
+            for (String element : c) {
+                pw.println(element);
+            }
+        } catch (IOException e) {
+            log.error(e);
+            super.removeAll(c);
+            int index = size() - c.size() - 1;
+            index = index < 0 ? 0 : index;
+            tryToRewriteFileContent(index);
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public boolean addAll(int index, Collection<? extends String> c) {
+        int lastSize = size();
+        if (!super.addAll(index, c)) {
+            return false;
+        }
+        tryToRewriteFileContent(index);
+        return (lastSize + c.size()) == size();
     }
 
     private void loadFileContent(File file) {
         try (BufferedReader br = new BufferedReader(new FileReader(file))) {
             String line;
             while ((line = br.readLine()) != null) {
                 if (!line.startsWith("#")) {
-                    list.add(line);
+                    super.add(line);
                 }
             }
         } catch (IOException e) {
-            throw new IllegalStateException("Cannot read file " + file.getPath(), e);
+            throw new IllegalStateException("Cannot access file " + file.getPath(), e);
+        }
+    }
+
+    private void tryToRewriteFileContent(int fromIndex) {
+        try (RandomAccessFile randomAccessFile = new RandomAccessFile(file.getAbsolutePath(), "rw")){
+            int characterCount = 0;
+            if (randomAccessFile.length() > 0 && fromIndex > 0) {
+                for (int i = 0; i < fromIndex; i++) {
+                    characterCount += get(i).length() + LINE_SEPARATOR_LENGTH;
+                }
+            }
+            randomAccessFile.setLength(characterCount);
+            randomAccessFile.seek(characterCount);
+
+            for (int i = fromIndex; i < size(); i++) {
+                randomAccessFile.writeBytes(get(i) + Utils.LINE_SEPARATOR);
+            }
+        } catch (IOException e) {
+            log.error(e);
+            super.clear();
+            loadFileContent(file);
```

Comment:
If some element fails you clear whole parent list? and then load to it everything that have been successfully written to file?

take for instance: memory: [A, B, C], file: [A, B, C]
set(1, X);
memory: [A, X, C] file: start writing [A, X] then fail -> clear memory so: [] then load everything from file so:
memory: [A, X], file: [A, X]

----
so user called set(1, X) on list [A, B, C] and the result is [A, X] ? IMHO this is a really big error.

You should take care of alignment memory vs. file on the level of individual methods - even if this catch block would have made sense then clearing super list and loading it again in set() operation which concerns only one element is really redundant.